dev/NumberBox/NumberBox.cpp

```diff
@@ -0,0 +1,555 @@
+            StepValue(false);
+        }
+    }
+}
+
+// Increments or decrements value by StepFrequency, wrapping if necessary
+void NumberBox::StepValue(bool isPositive)
+{
+    // Validating input before and after
+    ValidateInput();
+    double oldVal = Value();
+    double newVal = oldVal;
+
+    if (isPositive)
+    {
+        newVal += StepFrequency();
+    }
+    else
+    {
+        newVal -= StepFrequency();
+    }
+
+    // MinMaxMode Wrapping
+    if (MinMaxMode() == winrt::NumberBoxMinMaxMode::WrapEnabled && GetBoundState(newVal) != BoundState::InBounds)
+    {
+        while (newVal > MaxValue())
+        {
+            newVal = MinValue() + (newVal - MaxValue()) - 1;
+        }
+        while (newVal < MinValue())
+        {
+            newVal = MaxValue() - abs(newVal - MinValue()) + 1;
+        }
+    }
+    // Input Overwriting - Coerce to min or max
+    else if (BasicValidationMode() == winrt::NumberBoxBasicValidationMode::InvalidInputOverwritten && GetBoundState(newVal) != BoundState::InBounds)
+    {
+        if (newVal > MaxValue() && (MinMaxMode() == winrt::NumberBoxMinMaxMode::MaxEnabled || MinMaxMode() == winrt::NumberBoxMinMaxMode::MinAndMaxEnabled))
+        {
+            newVal = MaxValue();
+        }
+        else if (newVal < MinValue() && (MinMaxMode() == winrt::NumberBoxMinMaxMode::MinEnabled || MinMaxMode() == winrt::NumberBoxMinMaxMode::MinAndMaxEnabled))
+        {
+            newVal = MinValue();
+        }
+    }
+    
+    // Safeguard for floating point imprecision errors
+    int StepFreqSigDigits = ComputePrecisionRounderSigDigits(newVal);
+    m_stepPrecisionRounder.SignificantDigits(StepFreqSigDigits);
+    newVal = m_stepPrecisionRounder.RoundDouble(newVal);
+
+    // Update Text and Revalidate new value
+    Value(newVal);
+    UpdateTextToValue();
+    ValidateInput();
+}
+
+
+// Check if text resembles formulaic input to determine if parser should be executed
+bool NumberBox::IsFormulaic(const winrt::hstring& in)
+{
+    std::wstring input(in);
+    std::wregex formula(L"^([0-9()\\s]*[+-/*^%]+[0-9()\\s]*)+$");
+    std::wregex negval(L"^-([0-9])+(.([0-9])+)?$");
+    return (std::regex_match(input, formula) && !std::regex_match(input, negval));
+}
+
+// Computes the number of significant digits that precision rounder should use. This helps to prevent floating point imprecision errors. 
+int NumberBox::ComputePrecisionRounderSigDigits(double newVal)
+{
+    auto const oldVal = Value();
+    auto const stepFreq = StepFrequency();
+
+    // Run formatter on both values to discard trailing and leading 0's.
+    std::wstring formattedVal(m_stepPrecisionFormatter.Format(oldVal));
+    std::wstring formattedStep(m_stepPrecisionFormatter.Format(stepFreq));
+    std::wstring formattedNew(m_stepPrecisionFormatter.Format(newVal));
+
+    // Get size of only decimal portion of both old numbers. 
+    int oldValSig = static_cast<int>(formattedVal.substr(formattedVal.find_first_of('.') + 1).size());
```

Comment:
Checking for period is dangerous since it's not always the decimal separator. I'm not sure I understand the role of this function, can you elaborate?

Reply:
This function helps deal with minor precision errors in hopefully the least intrusive way possible by finding out what the true number of significant digits the result of the calculation will have is so that it can be plugged into a rounder that will correct that precision error.  There might be a way easier way to solve that issue that I've blanked on but this way seems pretty reliable.

You're right about the decimal separator, that's an oversight that should be easy to fix.